<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Изучение-таблиц" data-toc-modified-id="Изучение-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Изучение таблиц</a></span></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Сколько-книг-вышло-после-1-января-2000-года?" data-toc-modified-id="Сколько-книг-вышло-после-1-января-2000-года?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Сколько книг вышло после 1 января 2000 года?</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li></ul></div>

<a class="anchor" id="0bullet"></a>

**<font size="6" color='red' > Анализ базы данных при помощи SQL </font>**  



**<font color='red' > Цель исследования </font>**   - проанализировать базу данных, состоящую из книг, авторов, издателей, а также пользовательские обзоры книг и на их основе сформулировать ценностное предложение для нового продукта.

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Задания**  

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Изучение таблиц

In [2]:
# импортируем библиотеки

import pandas as pd 
from sqlalchemy import create_engine 

In [3]:
# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
 db_config['pwd'], 
 db_config['host'], 
 db_config['port'], 
 db_config['db']) 

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# добавим красоты в оформление проекта
# а именно - возможность выделять текст разными цветами

from IPython.core.display import display, HTML

display(HTML("<font color='blue'>синий </font>"))
display(HTML("<font color='maroon'>коричневый </font>"))
display(HTML("<font color='olive'> оливковый </font>"))
display(HTML("<font color='navy'>темно-синий </font>"))
display(HTML("<font color='purple'> фиолетовый </font>"))
display(HTML("<font color='teal'>бирюзовый </font>"))
display(HTML("<font color='gray'>серый </font>"))
display(HTML("<font color='lime'>светло-зеленый </font>"))
display(HTML("<font color='green'>зеленый </font>"))
display(HTML("<font color='yellow'>желтый </font>"))
display(HTML("<font color='FUCHSIA'> розовый </font>"))
display(HTML("<font color='AQUA'> голубой </font>"))

**<font size="5" color='teal' >  Таблица books </font>**   

In [4]:
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


**<font size="5" color='teal' >  Таблица authors </font>** 

In [6]:
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


**<font size="5" color='teal' >  Таблица publishers </font>**  


In [8]:
query = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


**<font size="5" color='teal' >  Таблица ratings </font>**  

In [10]:
 query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


**<font size="5" color='teal' >  Таблица reviews </font>**  

In [12]:
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


# Ответы на вопросы

## Сколько книг вышло после 1 января 2000 года?

* [В начало](#0bullet)

In [13]:
query = '''
            SELECT COUNT(book_id) count_of_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query, con = engine)

count_of_books
0             819

**<font size="3" color='blue' >  Вывод </font>**  

После 1 января 2000 года было выпущено 819 книг.

## Для каждой книги посчитайте количество обзоров и среднюю оценку

* [В начало](#0bullet)

In [15]:
query = '''
            SELECT b.book_id, rat.avg_rating, rev.review_count
            FROM books b
            LEFT JOIN (SELECT book_id, AVG(rating) avg_rating FROM ratings GROUP BY book_id) rat ON b.book_id=rat.book_id
            LEFT JOIN (SELECT book_id, COUNT(*) review_count FROM reviews GROUP BY book_id) rev ON b.book_id=rev.book_id
        '''
pd.io.sql.read_sql(query, con = engine)

book_id  avg_rating  review_count
0        652    4.500000           2.0
1        273    4.500000           2.0
2         51    4.250000           5.0
3        951    4.000000           2.0
4        839    4.285714           4.0
..       ...         ...           ...
995       64    4.230769           4.0
996       55    5.000000           2.0
997      148    3.428571           3.0
998      790    3.500000           2.0
999      828    3.000000           2.0

[1000 rows x 3 columns]

**<font size="3" color='blue' >  Вывод </font>**  

Как предварительно видно по данным, для большинства книг было написано несколько обзоров и средние оценки варьируются.

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

* [В начало](#0bullet)

In [16]:
query = '''
            SELECT b.publisher_id, p.publisher, COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

**<font size="3" color='blue' >  Вывод </font>** 

Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. Это около 4% от всех книг в данных.

британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном. Главной заслугой издательства считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам. Похоже, это издательство действительно достойно внимания.

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [4]:
query = '''
SELECT
    authors.author AS автор,
    AVG(ratings.rating) AS средний_рейтинг
    FROM ratings
    JOIN books ON books.book_id = ratings.book_id
    JOIN authors ON books.author_id = authors.author_id
    WHERE ratings.book_id IN (
       SELECT ratings.book_id
       FROM ratings
       GROUP BY ratings.book_id
       HAVING COUNT(ratings.book_id) >= 50)
    GROUP BY authors.author_id
    ORDER BY AVG(ratings.rating) DESC
    LIMIT 1'''
data = pd.io.sql.read_sql(query, con = engine)
data

автор  средний_рейтинг
0  J.K. Rowling/Mary GrandPré         4.287097

**<font size="3" color='blue' >  Вывод </font>**   

Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Пттере. Тут она представлена в соавторстве с Мэри Гранпрэ(иллюстратором). Что не удивительно, потому что книги о Гарри Поттере одни из самых популярных в мире среди всех возрастных категорий.

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

* [В начало](#0bullet)

In [5]:
query = '''

  SELECT AVG (reviews_count) as avg_reviews_count
  FROM ( 
        SELECT username,
        COUNT(review_id) as reviews_count
        FROM reviews
        GROUP BY username
        HAVING username IN (
                            SELECT username
                            FROM ratings
                            GROUP BY username
                            HAVING COUNT (rating_id) > 50
                            )
                            
                            
        ) as reviews
  
  '''



In [7]:
print('Читатели, которые поставили больше 50 оценок, в среднем пишут по {:.0f} обзора' \
      .format(pd.io.sql.read_sql(query, con = engine).iloc[0,0].round()))

Читатели, которые поставили больше 50 оценок, в среднем пишут по 24 обзора


**<font size="3" color='blue' >  Вывод </font>**  

В нашем распоряжении было 5 таблиц:  

* **books** - содержит 1000 строк и 6 столбцов, содержит данные о книгах;
* **authors** - содержит 636 строк и 2 столбца, содержит данные об авторах;
* **publishers** - содержит 340 строк, 2 столбца, содержит данные об издательствах;
* **rating** - содержит 6456 строк, 4 столбца, содержит данные о пользовательских обзорах;
* **reviews** - содержит 2793 строки, 4 столбца, содержит данные о пользовательских обзорах.


В ходе работы были даны ответы на 5 поставленных вопросов:  

- Cколько книг вышло после 1 января 2000 года? с 1 января 2000 года, в базе хранится информация о 819 книгах  

- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — таким образом исключим из анализа брошюры. Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. Это около 4% от всех книг в данных. Британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном. Главной заслугой издательства считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам. Похоже, это издательство действительно достойно внимания.  

- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.  Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Пттере. Тут она представлена в соавторстве с Мэри Гранпрэ(иллюстратором). Что не удивительно, потому что книги о Гарри Поттере одни из самых популярных в мире среди всех возрастных категорий.

- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.Таким образом среднее количество обзоров от активных пользователей. При подсчете было выявлено - 24 обзора.  

* [В начало](#0bullet)